lib/puppet/pops/types/p_sem_ver_range_type.rb

```diff
@@ -0,0 +1,155 @@
+  #
+  # @api public
+  def self.include?(range, version)
+    case version
+    when Semantic::Version
+      range.include?(version)
+    when String
+      begin
+        range.include?(Semantic::Version.parse(version))
+      rescue Semantic::Version::ValidationFailure
+        false
+      end
+    else
+      false
+    end
+  end
+
+  # Checks if range _a_ is a sub-range of (i.e. completely covered by) range _b_
+  # @param a [Semantic::VersionRange] the first range
+  # @param b [Semantic::VersionRange] the second range
+  #
+  # @return [Boolean] `true` if _a_ is completely covered by _b_
+  def self.covered_by?(a, b)
+    b.begin <= a.begin && (b.end > a.end || b.end == a.end && (!b.exclude_end? || a.exclude_end?))
+  end
+
+  # Merge two ranges so that the result matches all versions matched by both. A merge
+  # is only possible when the ranges are either adjacent or have an overlap.
+  #
+  # @param a [Semantic::VersionRange] the first range
+  # @param b [Semantic::VersionRange] the second range
+  # @return [Semantic::VersionRange,nil] the result of the merge
+  #
+  # @api public
+  def self.merge(a, b)
+    if a.include?(b.begin) || b.include?(a.begin)
+      max = [a.end, b.end].max
+      exclude_end = false
+      if a.exclude_end?
+        exclude_end = max == a.end && (max > b.end || b.exclude_end?)
+      elsif b.exclude_end?
+        exclude_end = max == b.end && (max > a.end || a.exclude_end?)
+      end
+      Semantic::VersionRange.new([a.begin, b.begin].min, max, exclude_end)
+    elsif a.exclude_end? && a.end == b.begin
+      # Adjacent, a before b
+      Semantic::VersionRange.new(a.begin, b.end, b.exclude_end?)
+    elsif b.exclude_end? && b.end == a.begin
+      # Adjacent, b before a
+      Semantic::VersionRange.new(b.begin, a.end, a.exclude_end?)
+    else
+      # No overlap
+      nil
+    end
+  end
+
+  def instance?(o, guard = nil)
+    o.is_a?(Semantic::VersionRange)
+  end
+
+  def eql?(o)
+    self.class == o.class
+  end
+
+  def hash?
+    super ^ @version_range.hash
+  end
+
+  def self.new_function(_, loader)
+    range_expr = "\\A#{range_pattern}\\Z"
+    @@new_function ||= Puppet::Functions.create_loaded_function(:new_VersionRange, loader) do
+      # Constructs a regexp pattern that matches a VersionRange as specified by
+      #
+      # https://github.com/npm/node-semver#range-grammar
+      #
+      # The logical or || operator is not implemented since it effectively builds
+      # an array of ranges that may be disparate. The {{Semantic::VersionRange}} inherits
+      # from the standard ruby range. It must be possible to describe that range in terms
+      # of min, max, and exclude max.
+      #
+      # The Puppet Version type is parameterized and accepts multiple ranges so creating such
+      # constraints is still possible. It will just require several parameters rather than one
+      # parameter containing the '||' operator.
+      #
```

Comment:
Before we move this into the `new` function I think the documentation already present there must be reorganized and have a structure that will allow easy overview of what types that can be created and how. At present it's a just a long list of "conversions" with no index, and just patching more text onto that doesn't strike me as a solution that will be helpful to the user. Reorging and preparing that doc to be extended is beyond the scope of this PR. I therefor suggest that we leave the doc for the SemVer and SemVerRange where it is right now and instead create a separate ticket on how to improve the documentation of the `new` function to make it more extendable.

Side note:
I think the term "conversion" is fine when there's one argument, or perhaps one argument with an additional argument that says how the first argument should be interpreted (like Integer('0xFF', 16)) but it doesn't fit the bill for expressions like these:
```
$v1 = Version.new(1, 2, 3)
$v2 = Version.new(3, 2, 0, 'beta')
$range = VersionRange.new($v1, $v2, false)
```